### PR TITLE
Add CodeBuild CF and buildspec

### PR DESCRIPTION
**Pull request by prateekchaudhry**
### Summary

This PR adds CodeBuild Cloudformation template and corresponding buildspec file to the repo.
* These CodeBuild projects will trigger builds whenever there is a PR created or updated to the main repo. The builds will create artifacts, which will be stored in an S3 bucket - required to pre-exist and required as an input parameter by the CF template. Only a list of authorized Github users can trigger these builds (see FilterGroups in the template) which can be edited later by aws-cli or aws-sdk as required.
* Buildspec contains the required commands to install golang in a CodeBuild container and then build ECS Agent with the ```dockerfree-agent-image``` target. It also differentiates between build architecture types and names the build artifacts accordingly, and pass them towards S3 bucket to a path referenced via SHA of the commit that triggered the build

### Testing

New tests cover the changes:  No tests added, agent artifacts created by this method were tested with functional tests

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: build-infrastructure/codebuild-devbuild-stack.yml [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+      Artifacts:
+        Location: !Ref BuildBucketName
+        NamespaceType: NONE
+        OverrideArtifactName: true
+        Packaging: NONE
+        Path: development
+        Type: S3
+      BadgeEnabled: false
+      ConcurrentBuildLimit: 10
+      Description: A sample agent build on pr
```
Comment (fierlion) [inline build-infrastructure/codebuild-devbuild-stack.yml:34]: we'll want to update this description once we're ready to synthesize this for production.

File: build-infrastructure/codebuild-devbuild-stack.yml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        # This allow list can be modified using aws-cli or aws-sdk
+        # CodeBuild also supports pattern matches using regex, but this is not useful for listing different Github IDs
+        # so they have to be listed separately
+        FilterGroups:
+          - - Type: EVENT
+              Pattern: 'PULL_REQUEST_CREATED,PULL_REQUEST_UPDATED,PULL_REQUEST_REOPENED'
+            - Type: BASE_REF
+              Pattern: !Sub '^${GithubBranchName}$'
+            - Type: ACTOR_ACCOUNT_ID
+              Pattern: '5080306' # Prateek
```
Comment (fierlion) [inline build-infrastructure/codebuild-devbuild-stack.yml:61]: nit add github username here (and below) in comment so the reference is clear

File: buildspecs/build.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+           x86_64)
+             architecture="amd64"
+           ;;
+           aarch64)
+             architecture="arm64"
+           ;;
+         esac
+
+       # Tested with GOVERSION 1.15.x and 1.17.x
+       - GOVERSION="1.15.15"
```
Comment (fierlion) [inline buildspecs/build.yml:23]: is go 1.17 available -- can we default to that?

**Review by fierlion — APPROVED** [review]

File: build-infrastructure/codebuild-devbuild-stack.yml (hunk 1 above)
Comment (prateekchaudhry) [reply build-infrastructure/codebuild-devbuild-stack.yml:34]: Yeah, i missed updating this one, will do that

File: buildspecs/build.yml (hunk 3 above)
Comment (prateekchaudhry) [reply buildspecs/build.yml:23]: Yep, will update to 1.17

Comment (prateekchaudhry) [reply buildspecs/build.yml]: Reading from GO_VERSION in the repo now, pushed changes to address this

File: buildspecs/build.yml [hunk 4]
```diff
@@ -1,20 +1,66 @@
...
+       - architecture=""
+       # Same buildspec for different architectures - detect the architecture here and rename the artifacts accordingly
+       - case $(uname -m) in
+           x86_64)
+             architecture="amd64"
+           ;;
+           aarch64)
+             architecture="arm64"
+           ;;
+         esac
```
Comment (YashdalfTheGray) [inline buildspecs/build.yml:20]: Noice! 👌

**Pull request merged**